File: ios/Shared/SlabRenderer.swift
import CoreGraphics
import CoreText
import UIKit

/// Everything the renderer needs, with no dependency on how it was stored.
struct SlabModel: Equatable {
    var title: String
    var slot: String
    var accent: Accent
    var year: Int
    var today: DayDate
    var doneDaysOfYear: Set<Int>
    var streak: Int
    var totalDone: Int
    var remaining: Int
    var doneToday: Bool
    var mood: Mood = .awake
}

/// One text style: a face at a size, a colour, and tracking measured in ems.
struct TypeSpec {
    let font: UIFont
    let color: ARGB
    let kern: CGFloat

    init(_ size: CGFloat, _ color: ARGB, _ font: UIFont, _ tracking: CGFloat) {
        self.font = font
        self.color = color
        self.kern = tracking * size
    }
}

/**
 Draws the card: a flat colour block carrying one year of days.

 The same routine backs the home-screen widget and every card inside the app,
 so the two can never drift apart. Marking today inverts the block to ink and
 the grid to the ritual's colour — that flip is the whole visual reward.

 Sizes arrive in points rather than pixels, so `density` is 1 on iOS where the
 Android build passes the screen's dp scale; the numbers below are the same
 numbers either way.
 */
enum SlabRenderer {

    struct Config {
        var density: CGFloat = 1
        var header = true
        var footer = true
        var action = false
        var cat = true
        var quarterRuler = false
        var corner: CGFloat = 24
        var pad: CGFloat = 17
        var fillBackground = true
    }

    static let actionHeight: CGFloat = 34

    static func render(size: CGSize, model: SlabModel, config: Config = Config()) -> UIImage {
        let w = max(size.width, 1)
        let h = max(size.height, 1)
        return UIGraphicsImageRenderer(size: CGSize(width: w, height: h)).image { context in
            draw(in: context.cgContext, width: w, height: h, model: model, config: config)
        }
    }

    static func draw(in ctx: CGContext, width w: CGFloat, height h: CGFloat,
                     model: SlabModel, config cfg: Config) {

        let d = cfg.density
        func dp(_ v: CGFloat) -> CGFloat { v * d }

        ctx.textMatrix = CGAffineTransform(scaleX: 1, y: -1)

        let pad = dp(cfg.pad)
        let corner = dp(cfg.corner)
        let card = CGRect(x: 0, y: 0, width: w, height: h)

        // Marked days flip the card to ink and light the grid in the ritual's
        // colour, so a glance across the home screen says what is still open.
        let lit = model.doneToday
        let block = lit ? Palette.ink : model.accent.block
        let onBlock = lit ? Palette.paper : model.accent.onBlock
        let markInk = lit ? model.accent.block : Palette.ink
        let missed = onBlock.withAlpha(52)
        let future = onBlock.withAlpha(23)

        if cfg.fillBackground {
            fill(ctx, rect: card, radius: corner, color: block)
        }

        // ── Layout ──────────────────────────────────────────────────────────
        let headerH = cfg.header ? dp(38) : 0
        let footerH = cfg.footer ? dp(22) : 0
        let actionH = cfg.action ? dp(actionHeight) : 0

        var bottom = h - pad
        if cfg.action || cfg.footer {
            bottom -= max(actionH, footerH)
        }

        let gridTopBound = pad + headerH + dp(6)
        let gridBottomBound = max(bottom - dp(10), gridTopBound + dp(14))

        let cols = GridGeo.colsFor(model.year)
        let gapRatio: CGFloat = 0.30
        let innerW = w - 2 * pad
        var cell = innerW / (CGFloat(cols) + gapRatio * CGFloat(cols - 1))
        var gap = cell * gapRatio
        var gridH = CGFloat(GridGeo.rows) * cell + CGFloat(GridGeo.rows - 1) * gap

        let zoneH = gridBottomBound - gridTopBound
        if gridH > zoneH {
            cell = zoneH / (CGFloat(GridGeo.rows) + gapRatio * CGFloat(GridGeo.rows - 1))
            gap = cell * gapRatio
            gridH = CGFloat(GridGeo.rows) * cell + CGFloat(GridGeo.rows - 1) * gap
        }
        let gridLeft = pad
        let gridTop = gridTopBound + max((zoneH - gridH) / 2, 0)

        let yearLen = GridGeo.lengthOf(model.year)
        let startOff = GridGeo.startOffset(model.year)
        let isCurrentYear = model.today.year == model.year
        let todayDoy = isCurrentYear ? model.today.dayOfYear : -1

        // ── Header ──────────────────────────────────────────────────────────
        if cfg.header {
            var textRight = w - pad

            if cfg.cat {
                let px = dp(0.96)
                let catW = Cat.widthFor(px)
                let catH = Cat.heightFor(px)
                let boxW = catW + dp(11)
                let boxH = catH + dp(9)
                let boxL = w - pad - boxW
                let box = CGRect(x: boxL, y: pad - dp(2), width: boxW, height: boxH)
                fill(ctx, rect: box, radius: dp(10), color: lit ? model.accent.block : Palette.ink)
                Cat.draw(
                    in: ctx,
                    left: box.midX - catW / 2,
                    top: box.midY - catH / 2,
                    px: px,
                    mood: model.mood
                )
                textRight = boxL - dp(10)
            } else {
                // No cat: the streak takes the corner instead.
                let num = TypeSpec(dp(17), onBlock, Fonts.extraBold(dp(17)), -0.02)
                text(ctx, "\(model.streak)", at: CGPoint(x: w - pad, y: pad + dp(15)), spec: num, align: .right)
                let cap = TypeSpec(dp(8), onBlock.withAlpha(150), Fonts.semiBold(dp(8)), 0.10)
                text(ctx, "DAY STREAK", at: CGPoint(x: w - pad, y: pad + dp(26)), spec: cap, align: .right)
                textRight = w - pad - dp(58)
            }

            let slotSpec = TypeSpec(dp(8.5), onBlock.withAlpha(150), Fonts.semiBold(dp(8.5)), 0.10)
            text(ctx, model.slot.uppercased(), at: CGPoint(x: pad, y: pad + dp(8)), spec: slotSpec)

            let nameSpec = TypeSpec(dp(17), onBlock, Fonts.extraBold(dp(17)), -0.02)
            let name = fit(model.title, spec: nameSpec, maxWidth: max(textRight - pad, dp(30)))
            text(ctx, name, at: CGPoint(x: pad, y: pad + dp(27)), spec: nameSpec)

            if cfg.cat {
                let streakSpec = TypeSpec(dp(8.5), onBlock.withAlpha(170), Fonts.semiBold(dp(8.5)), 0.10)
                text(ctx, "\(model.streak) DAY STREAK",
                     at: CGPoint(x: pad, y: pad + dp(37)), spec: streakSpec)
            }
        }

        // ── Quarter ruler ───────────────────────────────────────────────────
        if cfg.quarterRuler {
            let spec = TypeSpec(dp(8), onBlock.withAlpha(140), Fonts.semiBold(dp(8)), 0.10)
            for (month, label) in [(1, "JAN"), (4, "APR"), (7, "JUL"), (10, "OCT")] {
                let col = GridGeo.colOf(DayDate(year: model.year, month: month, day: 1))
                text(ctx, label,
                     at: CGPoint(x: gridLeft + CGFloat(col) * (cell + gap), y: gridTop - dp(5)),
                     spec: spec)
            }
        }

        // ── The field ───────────────────────────────────────────────────────
        let r = cell * 0.30
        for col in 0..<cols {
            for row in 0..<GridGeo.rows {
                let doy = col * GridGeo.rows + row - startOff + 1
                if doy < 1 || doy > yearLen { continue }
                let rect = CGRect(
                    x: gridLeft + CGFloat(col) * (cell + gap),
                    y: gridTop + CGFloat(row) * (cell + gap),
                    width: cell,
                    height: cell
                )

                let colour: ARGB
                if model.doneDaysOfYear.contains(doy) {
                    colour = markInk
                } else if isCurrentYear && doy > todayDoy {
                    colour = future
                } else {
                    colour = missed
                }
                fill(ctx, rect: rect, radius: r, color: colour)

                if doy == todayDoy {
                    // Today wears a ring whether or not it is filled.
                    let ring = rect.insetBy(dx: -dp(1.3), dy: -dp(1.3))
                    ctx.setStrokeColor(markInk.cgColor)
                    ctx.setLineWidth(max(dp(1.2), 1.3))
                    ctx.addPath(CGPath(roundedRect: ring, cornerWidth: r * 1.7, cornerHeight: r * 1.7, transform: nil))
                    ctx.strokePath()
                }
            }
        }

        // ── Footer ──────────────────────────────────────────────────────────
        if cfg.footer || cfg.action {
            let baseY = h - pad - dp(4)

            let litSpec = TypeSpec(dp(12), onBlock, Fonts.extraBold(dp(12)), 0)
            let litText = "\(model.totalDone) lit"
            text(ctx, litText, at: CGPoint(x: pad, y: baseY), spec: litSpec)
            var x = pad + width(litText, spec: litSpec) + dp(9)

            ctx.setFillColor(onBlock.withAlpha(90).cgColor)
            ctx.fillEllipse(in: CGRect(
                x: x + dp(2) - dp(2), y: baseY - dp(3.5) - dp(2),
                width: dp(4), height: dp(4)
            ))
            x += dp(13)

            let leftSpec = TypeSpec(dp(12), onBlock.withAlpha(160), Fonts.semiBold(dp(12)), 0)
            text(ctx, "\(model.remaining) to go", at: CGPoint(x: x, y: baseY), spec: leftSpec)

            if cfg.action {
                drawAction(ctx, w: w, h: h, pad: pad, model: model, lit: lit, d: d)
            }
        }
    }

    /// Where the mark-today control sits. The widget lays its button over this.
    static func actionRect(width w: CGFloat, height h: CGFloat,
                           model: SlabModel, config cfg: Config) -> CGRect {
        let d = cfg.density
        let pad = cfg.pad * d
        let hgt = actionHeight * d
        let lit = model.doneToday
        let label = lit ? "Done" : "Mark"
        let spec = TypeSpec(11.5 * d, lit ? Palette.ink : Palette.paper, Fonts.extraBold(11.5 * d), 0)
        let checkW = lit ? 15 * d : 0
        let wide = width(label, spec: spec) + 28 * d + checkW
        return CGRect(x: w - pad - wide, y: h - pad - hgt, width: wide, height: hgt)
    }

    private static func drawAction(_ ctx: CGContext, w: CGFloat, h: CGFloat, pad: CGFloat,
                                   model: SlabModel, lit: Bool, d: CGFloat) {
        func dp(_ v: CGFloat) -> CGFloat { v * d }
        let hgt = dp(actionHeight)
        let label = lit ? "Done" : "Mark"
        let spec = TypeSpec(dp(11.5), lit ? Palette.ink : Palette.paper, Fonts.extraBold(dp(11.5)), 0)
        let checkW = lit ? dp(15) : 0
        let labelW = width(label, spec: spec)
        let wide = labelW + dp(28) + checkW
        let right = w - pad
        let r = CGRect(x: right - wide, y: h - pad - hgt, width: wide, height: hgt)

        fill(ctx, rect: r, radius: hgt / 2, color: lit ? model.accent.block : Palette.ink)

        let textX = r.midX - labelW / 2 + checkW / 2
        text(ctx, label, at: CGPoint(x: textX, y: r.midY + dp(4)), spec: spec)

        if lit {
            // Hand-drawn check: no font dependency, crisp at any size.
            let cx = textX - dp(9)
            let cy = r.midY
            let s = dp(3.6)
            ctx.setStrokeColor(Palette.ink.cgColor)
            ctx.setLineWidth(max(dp(1.9), 2))
            ctx.setLineCap(.round)
            ctx.setLineJoin(.round)
            ctx.beginPath()
            ctx.move(to: CGPoint(x: cx - s, y: cy + s * 0.05))
            ctx.addLine(to: CGPoint(x: cx - s * 0.15, y: cy + s * 0.72))
            ctx.addLine(to: CGPoint(x: cx + s, y: cy - s * 0.72))
            ctx.strokePath()
        }
    }

    // ── Drawing primitives ──────────────────────────────────────────────────

    enum Align { case left, right, center }

    private static func fill(_ ctx: CGContext, rect: CGRect, radius: CGFloat, color: ARGB) {
        ctx.setFillColor(color.cgColor)
        ctx.addPath(CGPath(roundedRect: rect, cornerWidth: radius, cornerHeight: radius, transform: nil))
        ctx.fillPath()
    }

    private static func attributed(_ string: String, _ spec: TypeSpec) -> NSAttributedString {
        NSAttributedString(string: string, attributes: [
            .font: spec.font,
            .foregroundColor: spec.color.uiColor,
            .kern: spec.kern
        ])
    }

    static func width(_ string: String, spec: TypeSpec) -> CGFloat {
        let line = CTLineCreateWithAttributedString(attributed(string, spec))
        return CGFloat(CTLineGetTypographicBounds(line, nil, nil, nil))
    }

    /// Draws `string` with its baseline at `point`, the way Canvas.drawText does.
    static func text(_ ctx: CGContext, _ string: String, at point: CGPoint,
                     spec: TypeSpec, align: Align = .left) {
        let line = CTLineCreateWithAttributedString(attributed(string, spec))
        var x = point.x
        switch align {
        case .left: break
        case .right: x -= CGFloat(CTLineGetTypographicBounds(line, nil, nil, nil))
        case .center: x -= CGFloat(CTLineGetTypographicBounds(line, nil, nil, nil)) / 2
        }
        ctx.textPosition = CGPoint(x: x, y: point.y)
        CTLineDraw(line, ctx)
    }

    /// Truncates `string` to `maxWidth`, adding an ellipsis only when it overflows.
    static func fit(_ string: String, spec: TypeSpec, maxWidth: CGFloat) -> String {
        if width(string, spec: spec) <= maxWidth { return string }
        var end = string.count
        while end > 1 && width(String(string.prefix(end)) + "…", spec: spec) > maxWidth {
            end -= 1
        }
        return String(string.prefix(max(end, 1))) + "…"
    }
}

File: app/src/main/java/com/ayan/ritual/render/SlabRenderer.kt
package com.ayan.ritual.render

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import java.time.LocalDate

/** Everything the renderer needs, with no dependency on how it was stored. */
data class SlabModel(
    val title: String,
    val slot: String,
    val accent: Accent,
    val year: Int,
    val today: LocalDate,
    val doneDaysOfYear: Set<Int>,
    val streak: Int,
    val totalDone: Int,
    val remaining: Int,
    val doneToday: Boolean,
    val mood: Mood = Mood.AWAKE
)

/**
 * Draws the card: a flat colour block carrying one year of days.
 *
 * The same routine backs the home-screen widget and every card inside the app,
 * so the two can never drift apart. Marking today inverts the block to ink and
 * the grid to the ritual's colour — that flip is the whole visual reward.
 */
object SlabRenderer {

    data class Config(
        val density: Float,
        val header: Boolean = true,
        val footer: Boolean = true,
        val action: Boolean = false,
        val cat: Boolean = true,
        val quarterRuler: Boolean = false,
        val cornerDp: Float = 24f,
        val padDp: Float = 17f,
        val fillBackground: Boolean = true
    )

    const val ACTION_HEIGHT_DP = 34f

    fun render(widthPx: Int, heightPx: Int, model: SlabModel, cfg: Config): Bitmap {
        val bmp = Bitmap.createBitmap(
            widthPx.coerceAtLeast(1),
            heightPx.coerceAtLeast(1),
            Bitmap.Config.ARGB_8888
        )
        draw(Canvas(bmp), widthPx.toFloat(), heightPx.toFloat(), model, cfg)
        return bmp
    }

    fun draw(canvas: Canvas, w: Float, h: Float, model: SlabModel, cfg: Config) {
        val d = cfg.density
        fun dp(v: Float) = v * d

        val pad = dp(cfg.padDp)
        val corner = dp(cfg.cornerDp)
        val card = RectF(0f, 0f, w, h)
        val p = Paint(Paint.ANTI_ALIAS_FLAG)

        // Marked days flip the card to ink and light the grid in the ritual's
        // colour, so a glance across the home screen says what is still open.
        val lit = model.doneToday
        val block = if (lit) Palette.INK else model.accent.block
        val onBlock = if (lit) Palette.PAPER else model.accent.onBlock
        val markInk = if (lit) model.accent.block else Palette.INK
        val missed = onBlock.withAlpha(52)
        val future = onBlock.withAlpha(23)

        if (cfg.fillBackground) {
            p.color = block
            canvas.drawRoundRect(card, corner, corner, p)
        }

        // ── Layout ──────────────────────────────────────────────────────────
        val headerH = if (cfg.header) dp(38f) else 0f
        val footerH = if (cfg.footer) dp(22f) else 0f
        val actionH = if (cfg.action) dp(ACTION_HEIGHT_DP) else 0f

        var bottom = h - pad
        if (cfg.action || cfg.footer) {
            bottom -= maxOf(actionH, footerH)
        }

        val gridTopBound = pad + headerH + dp(6f)
        val gridBottomBound = (bottom - dp(10f)).coerceAtLeast(gridTopBound + dp(14f))

        val cols = GridGeo.colsFor(model.year)
        val gapRatio = 0.30f
        val innerW = w - 2f * pad
        var cell = innerW / (cols + gapRatio * (cols - 1))
        var gap = cell * gapRatio
        var gridH = GridGeo.ROWS * cell + (GridGeo.ROWS - 1) * gap

        val zoneH = gridBottomBound - gridTopBound
        if (gridH > zoneH) {
            cell = zoneH / (GridGeo.ROWS + gapRatio * (GridGeo.ROWS - 1))
            gap = cell * gapRatio
            gridH = GridGeo.ROWS * cell + (GridGeo.ROWS - 1) * gap
        }
        val gridW = cols * cell + (cols - 1) * gap
        val gridLeft = pad
        val gridTop = gridTopBound + ((zoneH - gridH) / 2f).coerceAtLeast(0f)

        val yearLen = GridGeo.lengthOf(model.year)
        val startOff = GridGeo.startOffset(model.year)
        val isCurrentYear = model.today.year == model.year
        val todayDoy = if (isCurrentYear) model.today.dayOfYear else -1

        // ── Header ──────────────────────────────────────────────────────────
        if (cfg.header) {
            var textRight = w - pad

            if (cfg.cat) {
                val px = dp(0.95f)
                val catW = Cat.widthFor(px)
                val catH = Cat.heightFor(px)
                val boxW = catW + dp(11f)
                val boxH = catH + dp(9f)
                val boxL = w - pad - boxW
                val box = RectF(boxL, pad - dp(2f), boxL + boxW, pad - dp(2f) + boxH)
                p.color = if (lit) model.accent.block else Palette.INK
                canvas.drawRoundRect(box, dp(10f), dp(10f), p)
                Cat.draw(
                    canvas,
                    box.centerX() - catW / 2f,
                    box.centerY() - catH / 2f,
                    px,
                    model.mood
                )
                textRight = boxL - dp(10f)
            } else {
                // No cat: the streak takes the corner instead.
                val num = paint(dp(17f), onBlock, Fonts.extraBold(), -0.02f).apply {
                    textAlign = Paint.Align.RIGHT
                }
                canvas.drawText(model.streak.toString(), w - pad, pad + dp(15f), num)
                val cap = paint(dp(8f), onBlock.withAlpha(150), Fonts.semiBold(), 0.10f).apply {
                    textAlign = Paint.Align.RIGHT
                }
                canvas.drawText("DAY STREAK", w - pad, pad + dp(26f), cap)
                textRight = w - pad - dp(58f)
            }

            val slotPaint = paint(dp(8.5f), onBlock.withAlpha(150), Fonts.semiBold(), 0.10f)
            canvas.drawText(model.slot.uppercase(), pad, pad + dp(8f), slotPaint)

            val namePaint = paint(dp(17f), onBlock, Fonts.extraBold(), -0.02f)
            val name = fit(namePaint, model.title, (textRight - pad).coerceAtLeast(dp(30f)))
            canvas.drawText(name, pad, pad + dp(27f), namePaint)

            if (cfg.cat) {
                val streakPaint = paint(dp(8.5f), onBlock.withAlpha(170), Fonts.semiBold(), 0.10f)
                canvas.drawText(
                    "${model.streak} DAY STREAK",
                    pad, pad + dp(37f), streakPaint
                )
            }
        }

        // ── Quarter ruler ───────────────────────────────────────────────────
        if (cfg.quarterRuler) {
            val rp = paint(dp(8f), onBlock.withAlpha(140), Fonts.semiBold(), 0.10f)
            val marks = listOf(1 to "JAN", 4 to "APR", 7 to "JUL", 10 to "OCT")
            for ((month, label) in marks) {
                val col = GridGeo.colOf(LocalDate.of(model.year, month, 1))
                canvas.drawText(label, gridLeft + col * (cell + gap), gridTop - dp(5f), rp)
            }
        }

        // ── The field ───────────────────────────────────────────────────────
        val r = cell * 0.30f
        val rect = RectF()
        for (col in 0 until cols) {
            for (row in 0 until GridGeo.ROWS) {
                val doy = col * GridGeo.ROWS + row - startOff + 1
                if (doy < 1 || doy > yearLen) continue
                val l = gridLeft + col * (cell + gap)
                val t = gridTop + row * (cell + gap)
                rect.set(l, t, l + cell, t + cell)

                p.color = when {
                    model.doneDaysOfYear.contains(doy) -> markInk
                    isCurrentYear && doy > todayDoy -> future
                    else -> missed
                }
                canvas.drawRoundRect(rect, r, r, p)

                if (doy == todayDoy) {
                    // Today wears a ring whether or not it is filled.
                    p.color = markInk
                    p.style = Paint.Style.STROKE
                    p.strokeWidth = maxOf(dp(1.2f), 1.3f)
                    val ring = RectF(rect).apply { inset(-dp(1.3f), -dp(1.3f)) }
                    canvas.drawRoundRect(ring, r * 1.7f, r * 1.7f, p)
                    p.style = Paint.Style.FILL
                }
            }
        }

        // ── Footer ──────────────────────────────────────────────────────────
        if (cfg.footer || cfg.action) {
            val baseY = h - pad - dp(4f)

            val litPaint = paint(dp(12f), onBlock, Fonts.extraBold(), 0f)
            val litText = "${model.totalDone} lit"
            canvas.drawText(litText, pad, baseY, litPaint)
            var x = pad + litPaint.measureText(litText) + dp(9f)

            p.color = onBlock.withAlpha(90)
            canvas.drawCircle(x + dp(2f), baseY - dp(3.5f), dp(2f), p)
            x += dp(13f)

            val leftPaint = paint(dp(12f), onBlock.withAlpha(160), Fonts.semiBold(), 0f)
            canvas.drawText("${model.remaining} to go", x, baseY, leftPaint)

            if (cfg.action) {
                drawAction(canvas, w, h, pad, model, lit, d)
            }
        }
    }

    /** The mark-today control. The widget lays a transparent hit target over it. */
    private fun drawAction(
        canvas: Canvas, w: Float, h: Float, pad: Float,
        model: SlabModel, lit: Boolean, d: Float
    ) {
        fun dp(v: Float) = v * d
        val p = Paint(Paint.ANTI_ALIAS_FLAG)
        val hgt = dp(ACTION_HEIGHT_DP)
        val label = if (lit) "Done" else "Mark"
        val labelPaint = paint(dp(11.5f), if (lit) Palette.INK else Palette.PAPER, Fonts.extraBold(), 0f)
        val checkW = if (lit) dp(15f) else 0f
        val wide = labelPaint.measureText(label) + dp(28f) + checkW
        val right = w - pad
        val r = RectF(right - wide, h - pad - hgt, right, h - pad)

        p.color = if (lit) model.accent.block else Palette.INK
        canvas.drawRoundRect(r, hgt / 2f, hgt / 2f, p)

        var textX = r.centerX() - labelPaint.measureText(label) / 2f + checkW / 2f
        canvas.drawText(label, textX, r.centerY() + dp(4f), labelPaint)

        if (lit) {
            // Hand-drawn check: no font dependency, crisp at any size.
            val cx = textX - dp(9f)
            val cy = r.centerY()
            val s = dp(3.6f)
            p.color = Palette.INK
            p.style = Paint.Style.STROKE
            p.strokeWidth = maxOf(dp(1.9f), 2f)
            p.strokeCap = Paint.Cap.ROUND
            p.strokeJoin = Paint.Join.ROUND
            val path = android.graphics.Path().apply {
                moveTo(cx - s, cy + s * 0.05f)
                lineTo(cx - s * 0.15f, cy + s * 0.72f)
                lineTo(cx + s, cy - s * 0.72f)
            }
            canvas.drawPath(path, p)
            p.style = Paint.Style.FILL
        }
    }

    fun paint(size: Float, color: Int, face: Typeface, tracking: Float): Paint =
        Paint(Paint.ANTI_ALIAS_FLAG).apply {
            textSize = size
            this.color = color
            typeface = face
            letterSpacing = tracking
            isSubpixelText = true
        }

    /** Truncates [text] to [maxWidth], adding an ellipsis only when it overflows. */
    fun fit(paint: Paint, text: String, maxWidth: Float): String {
        if (paint.measureText(text) <= maxWidth) return text
        var end = text.length
        while (end > 1 && paint.measureText(text.substring(0, end) + "…") > maxWidth) end--
        return text.substring(0, end.coerceAtLeast(1)) + "…"
    }
}
